massa-protocol-worker-2/src/handlers/operation_handler/retrieval.rs

```diff
@@ -0,0 +1,471 @@
+                        LruCache::new(
+                            NonZeroUsize::new(self.config.max_known_ops_size)
+                                .expect("max_known_ops_size in config must be > 0"),
+                        )
+                    });
+            for prefix in &op_batch {
+                known_ops.put(*prefix, ());
+            }
+        }
+
+        // filter out the operations that we already know about
+        {
+            let cache_read = self.cache.read();
+            op_batch.retain(|prefix| !cache_read.checked_operations_prefix.contains(prefix));
+        }
+
+        let mut ask_set = OperationPrefixIds::with_capacity(op_batch.len());
+        let mut future_set = OperationPrefixIds::with_capacity(op_batch.len());
+        // exactitude isn't important, we want to have a now for that function call
+        let now = Instant::now();
+        let mut count_reask = 0;
+        for op_id in op_batch {
+            let wish = match self.asked_operations.get_mut(&op_id) {
+                Some(wish) => {
+                    if wish.1.contains(peer_id) {
+                        continue; // already asked to the `peer_id`
+                    } else {
+                        Some(wish) // already asked but at someone else
+                    }
+                }
+                None => None,
+            };
+            if let Some(wish) = wish {
+                // Ask now if latest ask instant < now - operation_batch_proc_period
+                // otherwise add in future_set
+                if wish.0
+                    < now
+                        .checked_sub(self.config.operation_batch_proc_period.into())
+                        .ok_or(TimeError::TimeOverflowError)?
+                {
+                    count_reask += 1;
+                    ask_set.insert(op_id);
+                    wish.0 = now;
+                    wish.1.push(peer_id.clone());
+                } else {
+                    future_set.insert(op_id);
+                }
+            } else {
+                ask_set.insert(op_id);
+                self.asked_operations
+                    .put(op_id, (now, vec![peer_id.clone()]));
+            }
+        } // EndOf for op_id in op_batch:
+
+        if count_reask > 0 {
+            massa_trace!("re-ask operations.", { "count": count_reask });
+        }
+        if self.op_batch_buffer.len() < self.config.operation_batch_buffer_capacity
+            && !future_set.is_empty()
+        {
+            self.op_batch_buffer.push_back(OperationBatchItem {
+                instant: now
+                    .checked_add(self.config.operation_batch_proc_period.into())
+                    .ok_or(TimeError::TimeOverflowError)?,
+                peer_id: peer_id.clone(),
+                operations_prefix_ids: future_set,
+            });
+        }
+        if !ask_set.is_empty() {
+            {
+                let mut active_connections_write = self.active_connections.write();
```

Comment:
think we don't need .write() here

Reply:
Yes my bad in my head `send` still need a ref mut but we changed it. Done